torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1830,14 +1873,72 @@ def _local_pre_load_state_dict_hook(
                 f"the state_dict is {load_tensor.numel()}."
             )
             load_tensor = F.pad(load_tensor, [0, flat_param.num_padded])
-        state_dict[key] = load_tensor
+        state_dict[fqn] = load_tensor
 
     def _sharded_pre_load_state_dict_hook(
         self,
         state_dict: Dict[str, Any],
         prefix: str,
     ) -> None:
-        raise NotImplementedError("Will be implemented as part of https://github.com/pytorch/pytorch/issues/73518.")
+        """
+        The hook combines the unflattened, sharded parameters (ShardedTensor) to
+        a new FlatParameter and shards the new FlatParameter to the local chunk.
+        """
+        _replace_by_prefix(state_dict, prefix, prefix + f"{FSDP_WRAPPED_MODULE}.")
+        if self.module.no_params:
+            return
+
+        if not self.module.flat_param._is_sharded:
+            raise RuntimeError(
+                "load_sharded_state_dict can only be called when parameters "
+                "are flatten and sharded."
+            )
+
+        nonsharded_tensors = []
+        # TODO: Reduce the communication by using only one _all_gather_base to
+        # gather all the parameters in this layer. This can be achieved by
+        # concatenated all the local shards and then append the padding.
+        for module_name, _, param_name in self.module.flat_param._param_infos:
```

Comment:
could we create a Github issue regarding followup optimizations as we discussed offline, like coalesced all gather to reduce communication here?